### PR TITLE
travis tweaks to make sure the correct versions of python are installed

### PR DESCRIPTION
**Pull request by anderspapitto**
in order to make sure that the correct versions of python are installed on linux, set language=python in those configurations, while still manually installing python for the macos configurations

### CODE REVIEW

File: .travis/before_install.sh [hunk 1]
```diff
@@ -8,6 +8,9 @@ if [ ! -n "$NUMCORES" ]; then
 fi
 echo Using $NUMCORES cores
 
+echo ARST
```
Comment (anderspapitto) [inline .travis/before_install.sh]: I'll clean this up before pushing (and below), but I don't want to retrigger all the tests right now

File: .travis.yml [hunk 2]
```diff
@@ -1,32 +1,63 @@
...
+
+after_failure:
+  - ./.travis/after_failure.sh
 
 cache:
+  - timeout: 300
   - directories:
-    - $HOME/.build_cache
+    - $BUILD_CCACHE_DIR
+    - $BUILD_NINJA_DIR
```
Comment (bddppq) [inline .travis.yml]: remove this

File: .travis/before_install.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+echo NEIO
+python --version
+pip install virtualenv
+virtualenv -p "${PYTHON_VERSION}" "${HOME}/virtualenv"
+source "${HOME}/virtualenv/bin/activate"
+python --version
+
+pip install pytest-cov nbval
+
+if [[ $USE_NINJA == true ]]; then
```
Comment (bddppq) [inline .travis/before_install.sh]: remove this

File: .travis/before_install.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    pip install ninja
+fi
+
+# Update all existing python packages
```
Comment (bddppq) [inline .travis/before_install.sh]: Move this to right after virtualenv activation

File: .travis/before_install.sh [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  echo Unknown OS: $TRAVIS_OS_NAME
+  exit 1
+fi
+
+# TODO consider using "python3.6 -m venv"
+# which is recommended by python3.6 and may make some difference
+echo NEIO
+python --version
+pip install virtualenv
+virtualenv -p "${PYTHON_VERSION}" "${HOME}/virtualenv"
```
Comment (bddppq) [inline .travis/before_install.sh]: In travis linux, if language is set to python, then it has already set up a virtualenv.

File: .travis/before_install.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+echo ARST
+python --version
+
+# Install dependencies
+if [ "$TRAVIS_OS_NAME" == "linux" ]; then
+  sudo apt-get update
+  APT_INSTALL_CMD='sudo apt-get install -y --no-install-recommends'
+  $APT_INSTALL_CMD dos2unix
+
+  # Install protobuf
```
Comment (bddppq) [inline .travis/before_install.sh]: nit: wrap this protobuf installation into a function

File: .travis/before_install.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  pb_dir="~/.cache/pb"
+  mkdir -p "$pb_dir"
+  wget -qO- "https://github.com/google/protobuf/releases/download/v${PB_VERSION}/protobuf-${PB_VERSION}.tar.gz" | tar -xz -C "$pb_dir" --strip-components 1
+  ccache -z
+  cd "$pb_dir" && ./configure && make -j${NUMCORES} && make check && sudo make install && sudo ldconfig
+  ccache -s
+elif [ "$TRAVIS_OS_NAME" == "osx" ]; then
+  brew install ccache protobuf
+  if [ "${PYTHON_VERSION}" == "python3" ]; then
+    brew upgrade python
```
Comment (bddppq) [inline .travis/before_install.sh]: What's the reason of doing upgrade?

Comment (anderspapitto) [reply .travis/before_install.sh]: I took it from https://github.com/onnx/onnx-coreml/commit/f8af8753e266d0d02e3301e070333416e9942a5c, I believe it's what actually upgrades from python2 to python3, will confirm with @smessmer

File: .travis/before_install.sh (hunk 6 above)
Comment (anderspapitto) [reply .travis/before_install.sh]: ok, sure

File: .travis/before_install.sh (hunk 5 above)
Comment (anderspapitto) [reply .travis/before_install.sh]: ok, I'll try putting this under an IF

File: .travis/before_install.sh (hunk 3 above)
Comment (anderspapitto) [reply .travis/before_install.sh]: ++

File: .travis.yml (hunk 2 above)
Comment (anderspapitto) [reply .travis.yml]: ++

File: .travis/before_install.sh (hunk 7 above)
Comment (anderspapitto) [reply .travis/before_install.sh]: confirmed

File: .travis/before_install.sh [hunk 8]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+# Don't source setup.sh here, because the virtualenv might not be set up yet
+
+export NUMCORES=`grep -c ^processor /proc/cpuinfo`
+if [ ! -n "$NUMCORES" ]; then
```
Comment (bddppq) [inline .travis/before_install.sh]: nit: `-z`

File: .travis/before_install.sh [hunk 9]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+# Don't source setup.sh here, because the virtualenv might not be set up yet
+
+export NUMCORES=`grep -c ^processor /proc/cpuinfo`
```
Comment (bddppq) [inline .travis/before_install.sh]: nit: There is no need to do `export` here.

File: .travis/before_install.sh [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+# Install dependencies
+if [ "$TRAVIS_OS_NAME" == "linux" ]; then
+  sudo apt-get update
+  APT_INSTALL_CMD='sudo apt-get install -y --no-install-recommends'
+  $APT_INSTALL_CMD dos2unix
+
+  function install_protobuf() {
+      # Install protobuf
+      pb_dir="~/.cache/pb"
```
Comment (bddppq) [inline .travis/before_install.sh]: nit: local

**Pull request merged**